Refactor IP filter parsing code

sledgehammer999 opened the pull request:
This kind of refactoring was a minor desire of mine from the moment I had learnt about string_view (around 2017).
This has an average decrease in parsing time of ~15.50%

The last major refactoring of the parsing code was done by me in ~2017 in #5310.

Some measurements: **old vs new**

1. DAT file 1
529 ms
420 ms

2. DAT file 2
370 ms
314 ms

3. P2P file 1
740 ms
640 ms

4. P2P file 2
345 ms
301 ms

sledgehammer999 commented:
Does anyone build with mingw? It crashes for me totally inexplicably. It crashes the moment it is supposed to enter `parseTxtFilterFile()`. However, it doesn't even execute a print statement at the very top of the function.

sledgehammer999 commented:
Damn it. It seems `from_chars()` is only supported from [gcc 8.1+](https://gcc.gnu.org/onlinedocs/libstdc++/manual/status.html#status.iso.2017).
That's a fundamental part of this PR.

sledgehammer999 commented:
Is there a reason we keep Ubuntu 18.04 support? The Actions/Travis CI can work on newer Ubuntu and there is another Ubuntu LTS out there (20.04). Maybe drop 18.04?

Chocobo1 commented on src/base/bittorrent/filterparserthread.cpp:
```diff
@@ -37,75 +42,133 @@
+        // std::string_view description;
+    };
+
+    template <bool isDat> IPFilterEntry parseLine(std::string_view line)
```
I think the template parameter is overused. Looking at the code it just splits into two distinctive code path, you really are writing two functions instead of one.

Chocobo1 commented on src/base/bittorrent/filterparserthread.cpp:
```diff
@@ -37,75 +42,133 @@
+            // to parse possible IPv6 literals wrapped in []. These characters can appear in
+            // the "some organization" part too. It isn't worth implementing such complex parsing
+            // for a format that doesn't support IPv6 and there aren't any lists with IPv6 in them.
+            if (const auto colon = line.rfind(':');
```
It unclear what type the function returns, I wouldn't use `auto` here.

glassez replied:
>It unclear what type the function returns,

It's just one of standard iterator-related methods like (c/r)begin/(c/r)end etc.

But seems code in this line does nothing... There is empty body in `if`. Do I miss something?

sledgehammer999 replied:
> There is empty body in if. Do I miss something?

uhm this is an if with an init statement. This line is the initstatement and the next one the condition. The body follows. (Look at the open/close parentheses).

glassez replied:
>uhm this is an if with an init statement. This line is the initstatement and the next one the condition. The body follows. 

Ooops... I didn't notice that it's splitted into several lines.

Chocobo1 commented on src/base/bittorrent/filterparserthread.cpp:
```diff
@@ -37,75 +42,133 @@
 
-    const int BUFFER_SIZE = 2 * 1024 * 1024; // 2 MiB
-    const int MAX_LOGGED_ERRORS = 5;
+    constexpr int BUFFER_SIZE = 2 * 1024 * 1024; // 2 MiB
```
I would reserve `constexpr` for statements that really require compiler to evaluate at compile-time.

sledgehammer999 replied:
Why not? If something can be calculated at compile time, let's do it.

Chocobo1 replied:
https://stackoverflow.com/a/28821610 specifically about the "So, why do we need constexpr?" part.
We can doesn't always mean we should... and this is a personal preference, I don't insist that much to be a blocker.

glassez replied:
> Talking from experience, by reserving `constexpr` usage, it is easier to follow the code to the part that needs compile-time evaluation.
> 
> Side note, actually a lot of keywords such as `noexcept` declares intention of the programmer and IMO whether to use it depends on our use case at hand.

👍

glassez replied:
>If something can be calculated at compile time, let's do it.

This constant is calculated in compile time regardless of `constexpr` specifier.

sledgehammer999 replied:
I feel like this should be calculated at compile time (the intent). So I want to mark it as constexpr.

Chocobo1 commented on src/base/bittorrent/filterparserthread.cpp:
```diff
@@ -133,11 +196,8 @@ int FilterParserThread::parseDATFilterFile()
-    int offset = 0;
-    int start = 0;
-    int endOfLine = -1;
+    std::array<char, BUFFER_SIZE> buffer = {}; // seems a bit faster than QVector
```
2MB on stack!? That is just too big, most platforms only have ~1MB by default.

sledgehammer999 replied:
Hmm, could it be this the reason why mingw outright crashes? I'll investigate.
I suppose I should change it back to std::vector.

Chocobo1 commented on src/base/bittorrent/filterparserthread.cpp:
```diff
@@ -148,293 +208,111 @@ int FilterParserThread::parseDATFilterFile()
-            }
-
-            start = endOfLine;
+            const auto& [range, accessLevel] = parseLine<isDat>(line);
```
I hope we can write out the exact type instead of `auto`.

glassez replied:
>I hope we can write out the exact type instead of `auto`...

Sorry, cannot exactly understand what do you want. Do you suggest to use older approach (explicitly declare both variables and use `std::tie`)?
[Structured binding declaration](https://en.cppreference.com/w/cpp/language/structured_binding) is one of the unquestionable advantages of C++17 for me.

>Also it should be `const auto &[range, accessLevel]`

:+1:

Chocobo1 commented on src/base/bittorrent/filterparserthread.h:
```diff
@@ -54,8 +54,7 @@ class FilterParserThread final : public QThread
     int trim(char *const data, int start, int end);
-    int parseDATFilterFile();
-    int parseP2PFilterFile();
+    template <bool isDat> int parseTxtFilterFile();
```
Isn't it a `const` function?

sledgehammer999 replied:
It modifies `m_filter`

Chocobo1 commented on src/base/bittorrent/filterparserthread.cpp:
```diff
@@ -148,293 +208,111 @@ int FilterParserThread::parseDATFilterFile()
             try
             {
-                m_filter.add_rule(startAddr, endAddr, lt::ip_filter::blocked);
+                m_filter.add_rule(std::move(*startAddr), std::move(*endAddr), lt::ip_filter::blocked);
```
The `add_rule()` takes a const reference, so `std::move` isn't justified.
Personally, I don't like to use `std::move` without a good reason, it is hard to maintain.

sledgehammer999 replied:
Sorry, I was hasty at that point in time. I didn't investigate the add_rule() signature.

Chocobo1 commented on src/base/bittorrent/filterparserthread.cpp:
```diff
@@ -37,75 +42,133 @@
+            auto posIt  = std::find_if_not(lastptr, IPstr.data() + IPstr.size(),
+                                          [](unsigned char c){ return std::isspace(c); } );
+            int pos = 0;
+            if (posIt != IPstr.data() + IPstr.size())
```
Add parentheses to arithmetic operations: `(IPstr.data() + IPstr.size())`

glassez commented:
> Damn it. It seems `from_chars()` is only supported from [gcc 8.1+](https://gcc.gnu.org/onlinedocs/libstdc++/manual/status.html#status.iso.2017).

Cannot find anything about `from_chars()` in linked docs...
>Is there a reason we keep Ubuntu 18.04 support?

Maybe the user base? However, I do not know any statistics...
Another reason is that we are of the opinion not to stop supporting something as long as it is not too burdensome for us.
But if we are not talking about a large user base, and there is no easy way to do without `from_chars()` here, then this PR may be a reason to drop Ubuntu 18.04 support.

ArcticGems commented:
> 
> 
> > Damn it. It seems `from_chars()` is only supported from [gcc 8.1+](https://gcc.gnu.org/onlinedocs/libstdc++/manual/status.html#status.iso.2017).
> 
> Cannot find anything about `from_chars()` in linked docs...
> 

Maybe he means proposal 'P0067R5' =
http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0067r5.html

sledgehammer999 commented:
>Cannot find anything about from_chars() in linked docs...

I think it is under "Elementary string conversions"

>Maybe the user base? However, I do not know any statistics...
Another reason is that we are of the opinion not to stop supporting something as long as it is not too burdensome for us.
But if we are not talking about a large user base, and there is no easy way to do without from_chars() here, then this PR may be a reason to drop Ubuntu 18.04 support.

Looking at it objectively, and unfortunately for me, the user base for IPFilter isn't so big (IMO) that would warrant an OS drop...

sledgehammer999 commented:
Since, this is depending on dropping Ubuntu 18.04 support, I am going to close the PR. Unfortunately for me, it doesn't like a good idea to drop it for this particular PR.
If no one argues otherwise, I am going to close it in a few days.

>Does anyone build with mingw? It crashes for me totally inexplicably. 

This is due to using and std::array with 2MB size. Too much for the stack. Switching to  std::vector fixes the problem.

glassez commented:
>Since, this is depending on dropping Ubuntu 18.04 support, I am going to close the PR.

Why close? It can wait for better time to drop Ubuntu 18.04 support.

xavier2k6 commented:
> > Since, this is depending on dropping Ubuntu 18.04 support, I am going to close the PR.
> 
> Why close? It can wait for better time to drop Ubuntu 18.04 support.

Perhaps it can coincide with when it comes to implement Qt 6.x?

Ref.:
https://github.com/qbittorrent/qBittorrent/pull/13937#issuecomment-741615844

sledgehammer999 commented:
I've changed my mind. We are stretched thin*, and it doesn't make sense to support older versions of OS's that aren't in widespread use. Especially when it hinders our ability to use recent c++ standards and recent library versions. At the moment, such platforms are Linux and macOS. We already support their latest versions. Older versions are dropped when it doesn't suit us.

*supporting older versions means maintaining local VMs to test when things go wrong, trying to find a fix etc aka maintenance work. This consumes time that can be spent elsewhere.

In my recent v4.3.3 news entry I indicated that Ubuntu 18.04 might be dropped. I didn't notice any user objection till today.

So anyway, I pushed fixes to my PR. Please resume the review. Since this touches a sufficiently isolated module(IP parser) I'll leave it in master and not backport it for the next few releases.

Chocobo1 commented on src/base/bittorrent/filterparserthread.cpp:
```diff
@@ -37,255 +42,151 @@
+        {
+            entry.range = line.substr(0, firstComma);
+
+            auto posIt  = std::find_if_not((line.cbegin() + firstComma + 1), line.cend(),
```
seems this can be `const`.

Chocobo1 commented on src/base/bittorrent/filterparserthread.cpp:
```diff
@@ -37,255 +42,151 @@
+
+            auto posIt  = std::find_if_not((line.cbegin() + firstComma + 1), line.cend(),
+                                          [](unsigned char c){ return std::isspace(c); } );
+            int pos = 0;
```
I would write: 
```c++
const int pos = (posIt == line.cend()) ? 0 : (posIt - line.cbegin());
```

Chocobo1 commented on src/base/bittorrent/filterparserthread.cpp:
```diff
@@ -37,255 +42,151 @@
+        // the "some organization" part too. It isn't worth implementing such complex parsing
+        // for a format that doesn't support IPv6 and there aren't any lists with IPv6 in them.
+        if (const auto colon = line.rfind(':');
+            colon != line.npos)
```
IMO putting it on the same line of `if` is easier to avoid tripping over when reading it (at least for me):
```c++ 
if (const auto colon = line.rfind(':'); colon != line.npos)
```

Chocobo1 commented on src/base/bittorrent/filterparserthread.cpp:
```diff
@@ -37,255 +42,151 @@
+            entry.range = line.substr(colon + 1);
         else
-            address = lt::make_address(data, ec);
+            entry.range = line;
```
Also why not just:
```c++
{
    const auto colon = line.rfind(':');
    return {(colon != line.npos) ? line.substr(colon + 1): line}};
}
```

Chocobo1 commented on src/base/bittorrent/filterparserthread.cpp:
```diff
@@ -37,255 +42,151 @@
-    {
-        LogMsg(tr("I/O Error: Could not open IP filter file in read mode."), Log::CRITICAL);
-        return ruleCount;
+        return std::make_pair(start, end);
```
Same as above, you can omit `start, end` variables and return `{}` instead.

Chocobo1 commented on src/base/bittorrent/filterparserthread.cpp:
```diff
@@ -312,129 +210,111 @@ int FilterParserThread::parseP2PFilterFile()
-            lt::address startAddr;
-            int newStart = trim(buffer.data(), partsDelimiter + 1, delimIP - 1);
-            if (!parseIPAddress(buffer.data() + newStart, startAddr))
+            std::optional<lt::address> startAddr = parseIPAddress(startIP);
```
seems this can be const.

Chocobo1 commented on src/base/bittorrent/filterparserthread.cpp:
```diff
@@ -312,129 +210,111 @@ int FilterParserThread::parseP2PFilterFile()
-            lt::address endAddr;
-            newStart = trim(buffer.data(), delimIP + 1, endOfLine);
-            if (!parseIPAddress(buffer.data() + newStart, endAddr))
+            std::optional<lt::address> endAddr = parseIPAddress(endIP);
```
This one too.

Chocobo1 commented on src/base/bittorrent/filterparserthread.cpp:
```diff
@@ -312,129 +210,111 @@ int FilterParserThread::parseP2PFilterFile()
-            if (delimIP == -1)
+            const auto &[range, accessLevel] = isDat ? parseDATLine(line) : parseP2PLine(line);
+
+            if constexpr (isDat)
```
please add curly brackets.

Chocobo1 commented on src/base/bittorrent/filterparserthread.cpp:
```diff
@@ -37,255 +42,151 @@
-                start = endOfLine;
-                continue;
-            }
+        if (auto addr = parseIPv4Address(IPstr);
```
`const auto addr`

glassez replied:
Please make such `if` declaration one-liner whenever possible.

Chocobo1 reviewed (commented):
some style suggestions.

glassez commented on src/base/bittorrent/filterparserthread.cpp:
```diff
@@ -37,255 +42,151 @@
-            if (firstComma != -1)
-                findAndNullDelimiter(buffer.data(), ',', firstComma + 1, endOfLine);
+        if (octetIndex == 4)
+            return lt::address_v4(octets);
```
Why not put it in loop body? Then you can omit additional cheking in loop condition.

glassez commented on src/base/bittorrent/filterparserthread.cpp:
```diff
@@ -37,255 +42,151 @@
-        if ((bytesRead == 0) && (dataSize == 0))
-            break;
+        lt::address_v4::bytes_type octets = {};
+        unsigned char octetIndex = 0;
```
Declaring a variable as `unsigned char` (and not just `int`) may not be of any use in such a case. Also I would avoid unsigned types unless you sure it really required in some case.

glassez commented on src/base/bittorrent/filterparserthread.cpp:
```diff
@@ -630,7 +510,7 @@ void FilterParserThread::run()
     {
         // eMule DAT format
-        ruleCount = parseDATFilterFile();
+        ruleCount = parseTxtFilterFile<true>();
```
Using `bool` parameters causes such confusing code. It's better to use named constants (or enums):
```c++
        ruleCount = parseTxtFilterFile<Format::DAT>();
```

sledgehammer999 commented:
Can you guys look again at this PR? I have addressed some comments. But since this has been done long time ago, can you start fresh as if this was new?

glassez commented on src/base/bittorrent/filterparserthread.cpp:
```diff
@@ -37,255 +42,137 @@
-                    return true;
-            }
+        // Each line should follow this format:
+        // 001.009.096.105 - 001.009.096.105 , 000 , Some organization
```
Either this line describe format incorrectly or you don't take into account spaces so they go into parsed items, e.g. `entry.range` becomes `001.009.096.105 - 001.009.096.105 ` (with space in the end).
The same is about  `parseRange()`.

sledgehammer999 replied:
>becomes 001.009.096.105 - 001.009.096.105 (with space in the end).

`std::from_chars()` will stop at the first non-numeric character. So we don't care about trailing spaces.

>The same is about parseRange().

But we care about leading spaces and that function doesn't account for that.

glassez reviewed (commented):
Some CI tests fail.

xavier2k6 commented:
Also for consideration, findings of #14434

sledgehammer999 commented:
>Some CI tests fail.

It seems that msvc doesn't like iterators. `std::from_chars()` takes only `const char*` for first/last. A string_view's iterators seem to decay to that type in gcc/clang but not in msvc.
Or at least, that's what I think the error message means. Can you look at the error message too and confirm?

sledgehammer999 commented:
>It seems that msvc doesn't like iterators. std::from_chars() takes only const char* for first/last. A string_view's iterators seem to decay to that type in gcc/clang but not in msvc.
Or at least, that's what I think the error message means. Can you look at the error message too and confirm?

While googling I came to [this](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p2007r0.html#a-naive-approach):
>The above will work for some implementations and build modes but will fail for others. This is caused by the fact that even though some vendors implement `std::string_view::const_iterator` in terms of `const char*`, such code is not portable. The C++ standard specifies s`td::string_view::const_iterator` as implementation defined and only requires that the iterator’s `value_type` is `char`.

So I will abandon the iterator approach since it isn't portable and I will revert back to using `.data()` as per http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p2007r0.html#a-correct-approach

sledgehammer999 commented:
Current issues fixed.

Chocobo1 commented on src/base/bittorrent/filterparserthread.cpp:
```diff
@@ -28,7 +28,12 @@
 #include "filterparserthread.h"
 
+#include <algorithm>
+#include <charconv>
```
wrong order

Chocobo1 commented on src/base/bittorrent/filterparserthread.cpp:
```diff
@@ -37,75 +42,120 @@
+        // std::string_view description;
+    };
+
+    IPFilterEntry parseDATLine(std::string_view line)
```
if `line` doesn't change value through out this function, you should consider adding `const` to it.
The same to parameters of other functions.

sledgehammer999 replied:
ok, but does it makes sense to use `const` with a string_view?

Chocobo1 replied:
>but does it makes sense to use const with a string_view?

Of course, just like using `const` on any variable: if you don't intend to modify it further and wish to keep the initialized value/view.

Chocobo1 commented on src/base/bittorrent/filterparserthread.cpp:
```diff
@@ -37,75 +42,120 @@
-            unsigned char octetIndex = 0;
+            entry.range = line.substr(0, firstComma);
+            const auto pos  = std::find_if_not((line.data() + firstComma + 1), (line.data() + line.size()),
+                                          [](unsigned char c) { return std::isspace(c); } );
```
For consistency: `const unsigned char c`

Chocobo1 commented on src/base/bittorrent/filterparserthread.cpp:
```diff
@@ -37,75 +42,120 @@
+        int octetIndex = 0;
+
+        for (auto lastPos = IPstr.data();
+             lastPos < (IPstr.data() + IPstr.size());
```
I don't object this but there are other lines longer than this for-loop so you don't need to fold it.

Chocobo1 commented on src/base/bittorrent/filterparserthread.cpp:
```diff
@@ -37,75 +42,120 @@
+    std::optional<lt::address> parseIPAddress(std::string_view IPstr)
     {
-        IPv4Parser parser;
         lt::error_code ec;
```
I would move it lower to where it is used.

NotTsunami commented:
@sledgehammer999 Is this still being worked on?

Chocobo1 commented:
> @sledgehammer999 Is this still being worked on?

You can take over (by submitting another PR) if you are interested. Note that the comments on code are still expected to be addressed.

summerqB commented:
> @sledgehammer999 Is this still being worked on?

Yes. The milestone is being changed with every release. Nothing more 😆
